ansible_rulebook/job_template_runner.py

```diff
@@ -89,65 +90,118 @@ def _sslcontext(self) -> Union[bool, ssl.SSLContext]:
-                f"{organization} does not exist"
-            )
-        )
-
     async def run_job_template(
         self,
         name: str,
         organization: str,
         job_params: dict,
     ) -> dict:
-        job = await self.launch(name, organization, job_params)
+        obj = await self._get_template_obj(name, organization, "job_template")
+        if not obj:
+            raise JobTemplateNotFoundException(
+                (
+                    f"Job template {name} in organization "
+                    f"{organization} does not exist"
+                )
+            )
+        url = urljoin(self.host, obj["launch"])
+        job = await self._launch(job_params, url)
+        return await self._monitor_job(job["url"])
 
-        url = job["url"]
-        params = {}
+    async def run_workflow_job_template(
+        self,
+        name: str,
+        organization: str,
+        job_params: dict,
+    ) -> dict:
+        obj = await self._get_template_obj(
+            name, organization, "workflow_job_template"
+        )
+        if not obj:
+            raise WorkflowJobTemplateNotFoundException(
+                (
+                    f"Workflow template {name} in organization "
+                    f"{organization} does not exist"
+                )
+            )
+        url = urljoin(self.host, obj["launch"])
+        if not obj["ask_limit_on_launch"] and "limit" in job_params:
+            logger.warn(
+                "Workflow template %s does not accept limit, removing it", name
+            )
+            job_params.pop("limit")
+        if not obj["ask_variables_on_launch"] and "extra_vars" in job_params:
+            logger.warn(
+                "Workflow template %s does not accept extra vars, "
+                "removing it",
+                name,
+            )
+            job_params.pop("extra_vars")
+        job = await self._launch(job_params, url)
+        return await self._monitor_job(job["url"])
 
+    async def _monitor_job(self, url) -> dict:
         while True:
             # fetch and process job status
-            json_body = await self._get_page(url, params)
-            job_status = json_body["status"]
-            if job_status in self.JOB_COMPLETION_STATUSES:
+            json_body = await self._get_page(url, {})
+            if json_body["status"] in self.JOB_COMPLETION_STATUSES:
                 return json_body
 
             await asyncio.sleep(self.refresh_delay)
 
-    async def launch(
-        self, name: str, organization: str, job_params: dict
-    ) -> dict:
-        jt_id = await self._get_job_template_id(name, organization)
-        url = urljoin(self.host, f"{self.JOB_TEMPLATE_SLUG}/{jt_id}/launch/")
-
+    async def _launch(self, job_params: dict, url: str) -> dict:
+        body = None
         try:
             async with self._session.post(
-                url, json=job_params, ssl=self._sslcontext
+                url,
```

Comment:
You can call directly to json() method, and no need to raise the exception later. The print of the error is redundant, because the exception contains the body. If we want to distinguish the error we can capture the different exceptions. 

I suggest:

```
        try:
            async with self._session.post(
                url,
                json=job_params,
                ssl=self._sslcontext,
                raise_for_status=True,
            ) as post_response:
                body = await post_response.json()
                return body
        except aiohttp.ClientResponseError as e:
            logger.error("Response error: %s", str(e))
            raise ControllerApiException(str(e))   
        except aiohttp.ClientError as e:
            logger.error("Error connecting to controller %s", str(e))
            raise ControllerApiException(str(e))
```

Reply:
@Alex-Izquierdo The issue is the body contains good error information like 

ERROR - Error {'extra_vars': ['Variables ansible_eda are not allowed on launch. Check the Prompt on Launch setting on the Workflow Job Template to include Extra Variables.']}

When you use the raise_for_status it doesn't parse the body since the body could be huge so we can't get the details.
Otherwise the error message would look like

ERROR - Error connecting to controller 400, message='Bad Request', url=URL('https://mycontroller/api/v2/workflow_job_templates/183/launch/')

In our case we know we are doing a POST and the response body is pretty small. so we have to do the raise_for_status=False and once we have the body we can call the raise_for_status. https://github.com/aio-libs/aiohttp/issues/3248

Reply:
You are right, I assumed wrongly the body it was in exception.